{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "20e7ad55-7569-4be5-b294-c801f72949f2",
   "metadata": {},
   "source": [
    "## Download data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "12f0b83e-7ce6-4a90-82f0-ae19bc49df25",
   "metadata": {},
   "outputs": [],
   "source": [
    "# wget \"https://ftp.ncbi.nlm.nih.gov/geo/series/GSE252nnn/GSE252432/suppl/GSE252432_RAW.tar\"import tarfile\n",
    "# wget \"ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE253nnn/GSE253173/suppl/GSE253173_single_cell_DREAM.h5ad.gz\"\n",
    "# gunzip GSE253173_single_cell_DREAM.h5ad.gz\n",
    "\n",
    "# import tarfile\n",
    "\n",
    "# # Path to .tar file\n",
    "# tar_path = 'GSE252432_RAW.tar'\n",
    "\n",
    "# # Extract to this directory\n",
    "# extract_path = './GSE252432_RAW'\n",
    "\n",
    "# # Open and extract\n",
    "# with tarfile.open(tar_path, 'r') as tar:\n",
    "#     tar.extractall(path=extract_path)\n",
    "\n",
    "# print(f\"Extracted to {extract_path}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8cb2a8b8-e9c0-4327-b72e-68acf9f7f443",
   "metadata": {},
   "source": [
    "## Read data and find matches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "82ae3782-2cc9-4fa1-af46-dcb1a0889a27",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode</th>\n",
       "      <th>is_cell</th>\n",
       "      <th>contig_id</th>\n",
       "      <th>high_confidence</th>\n",
       "      <th>length</th>\n",
       "      <th>chain</th>\n",
       "      <th>v_gene</th>\n",
       "      <th>d_gene</th>\n",
       "      <th>j_gene</th>\n",
       "      <th>c_gene</th>\n",
       "      <th>...</th>\n",
       "      <th>fwr3_nt</th>\n",
       "      <th>cdr3</th>\n",
       "      <th>cdr3_nt</th>\n",
       "      <th>fwr4</th>\n",
       "      <th>fwr4_nt</th>\n",
       "      <th>reads</th>\n",
       "      <th>umis</th>\n",
       "      <th>raw_clonotype_id</th>\n",
       "      <th>raw_consensus_id</th>\n",
       "      <th>exact_subclonotype_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCTGCATCCTTGC-1</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACCTGCATCCTTGC-1_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>479</td>\n",
       "      <td>TRB</td>\n",
       "      <td>TRBV5-5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ1-1</td>\n",
       "      <td>TRBC1</td>\n",
       "      <td>...</td>\n",
       "      <td>AGAGGAAGAGGAAACTTCCCTGATCGATTCTCAGCTCGCCAGTTCC...</td>\n",
       "      <td>CASSLGLNTEAFF</td>\n",
       "      <td>TGTGCCAGCAGCTTGGGGCTAAACACTGAAGCTTTCTTT</td>\n",
       "      <td>GQGTRLTVV</td>\n",
       "      <td>GGACAAGGCACCAGACTCACAGTTGTAG</td>\n",
       "      <td>2891</td>\n",
       "      <td>3</td>\n",
       "      <td>clonotype69</td>\n",
       "      <td>clonotype69_consensus_1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACGGGGTCCGTGAC-1</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACGGGGTCCGTGAC-1_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>568</td>\n",
       "      <td>TRA</td>\n",
       "      <td>TRAV19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRAJ44</td>\n",
       "      <td>TRAC</td>\n",
       "      <td>...</td>\n",
       "      <td>GAAATAAGTGGTCGGTATTCTTGGAACTTCCAGAAATCCACCAGTT...</td>\n",
       "      <td>CALSEDTGTASKLTF</td>\n",
       "      <td>TGTGCTCTGAGTGAGGATACCGGCACTGCCAGTAAACTCACCTTT</td>\n",
       "      <td>GTGTRLQVTL</td>\n",
       "      <td>GGGACTGGAACAAGACTTCAGGTCACGCTCG</td>\n",
       "      <td>801</td>\n",
       "      <td>2</td>\n",
       "      <td>clonotype195</td>\n",
       "      <td>clonotype195_consensus_2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACGGGGTCCGTGAC-1</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACGGGGTCCGTGAC-1_contig_2</td>\n",
       "      <td>True</td>\n",
       "      <td>517</td>\n",
       "      <td>TRB</td>\n",
       "      <td>TRBV14</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ2-1</td>\n",
       "      <td>TRBC2</td>\n",
       "      <td>...</td>\n",
       "      <td>CAGGATGAGTCCGGTATGCCCAACAATCGATTCTTAGCTGAAAGGA...</td>\n",
       "      <td>CASSLLTPTDDNEQFF</td>\n",
       "      <td>TGTGCCAGCAGCCTCCTTACACCGACGGACGACAATGAGCAGTTCTTC</td>\n",
       "      <td>GPGTRLTVL</td>\n",
       "      <td>GGGCCAGGGACACGGCTCACCGTGCTAG</td>\n",
       "      <td>23208</td>\n",
       "      <td>10</td>\n",
       "      <td>clonotype195</td>\n",
       "      <td>clonotype195_consensus_1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAAGTAGCAATCCGAT-1</td>\n",
       "      <td>True</td>\n",
       "      <td>AAAGTAGCAATCCGAT-1_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>502</td>\n",
       "      <td>TRB</td>\n",
       "      <td>TRBV7-9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ1-2</td>\n",
       "      <td>TRBC1</td>\n",
       "      <td>...</td>\n",
       "      <td>CTAGAAAAATCAAGGCTGCTCAGTGATCGGTTCTCTGCAGAGAGGC...</td>\n",
       "      <td>CASSLDGTNYGYTF</td>\n",
       "      <td>TGTGCCAGCAGCTTAGACGGCACTAACTATGGCTACACCTTC</td>\n",
       "      <td>GSGTRLTVV</td>\n",
       "      <td>GGTTCGGGGACCAGGTTAACCGTTGTAG</td>\n",
       "      <td>9293</td>\n",
       "      <td>3</td>\n",
       "      <td>clonotype16</td>\n",
       "      <td>clonotype16_consensus_1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAAGTAGTCACCCGAG-1</td>\n",
       "      <td>True</td>\n",
       "      <td>AAAGTAGTCACCCGAG-1_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>556</td>\n",
       "      <td>TRA</td>\n",
       "      <td>TRAV12-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRAJ40</td>\n",
       "      <td>TRAC</td>\n",
       "      <td>...</td>\n",
       "      <td>AATGAAGATGGAAGGTTTACAGCACAGCTCAATAGAGCCAGCCAGT...</td>\n",
       "      <td>CVVNMGGPGTYKYIF</td>\n",
       "      <td>TGTGTGGTGAACATGGGGGGCCCAGGAACCTACAAATACATCTTT</td>\n",
       "      <td>GTGTRLKVLA</td>\n",
       "      <td>GGAACAGGCACCAGGCTGAAGGTTTTAGCAA</td>\n",
       "      <td>6908</td>\n",
       "      <td>3</td>\n",
       "      <td>clonotype88</td>\n",
       "      <td>clonotype88_consensus_3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              barcode  is_cell                    contig_id  high_confidence  \\\n",
       "0  AAACCTGCATCCTTGC-1     True  AAACCTGCATCCTTGC-1_contig_1             True   \n",
       "1  AAACGGGGTCCGTGAC-1     True  AAACGGGGTCCGTGAC-1_contig_1             True   \n",
       "2  AAACGGGGTCCGTGAC-1     True  AAACGGGGTCCGTGAC-1_contig_2             True   \n",
       "3  AAAGTAGCAATCCGAT-1     True  AAAGTAGCAATCCGAT-1_contig_1             True   \n",
       "4  AAAGTAGTCACCCGAG-1     True  AAAGTAGTCACCCGAG-1_contig_1             True   \n",
       "\n",
       "   length chain    v_gene d_gene   j_gene c_gene  ...  \\\n",
       "0     479   TRB   TRBV5-5    NaN  TRBJ1-1  TRBC1  ...   \n",
       "1     568   TRA    TRAV19    NaN   TRAJ44   TRAC  ...   \n",
       "2     517   TRB    TRBV14    NaN  TRBJ2-1  TRBC2  ...   \n",
       "3     502   TRB   TRBV7-9    NaN  TRBJ1-2  TRBC1  ...   \n",
       "4     556   TRA  TRAV12-1    NaN   TRAJ40   TRAC  ...   \n",
       "\n",
       "                                             fwr3_nt              cdr3  \\\n",
       "0  AGAGGAAGAGGAAACTTCCCTGATCGATTCTCAGCTCGCCAGTTCC...     CASSLGLNTEAFF   \n",
       "1  GAAATAAGTGGTCGGTATTCTTGGAACTTCCAGAAATCCACCAGTT...   CALSEDTGTASKLTF   \n",
       "2  CAGGATGAGTCCGGTATGCCCAACAATCGATTCTTAGCTGAAAGGA...  CASSLLTPTDDNEQFF   \n",
       "3  CTAGAAAAATCAAGGCTGCTCAGTGATCGGTTCTCTGCAGAGAGGC...    CASSLDGTNYGYTF   \n",
       "4  AATGAAGATGGAAGGTTTACAGCACAGCTCAATAGAGCCAGCCAGT...   CVVNMGGPGTYKYIF   \n",
       "\n",
       "                                            cdr3_nt        fwr4  \\\n",
       "0           TGTGCCAGCAGCTTGGGGCTAAACACTGAAGCTTTCTTT   GQGTRLTVV   \n",
       "1     TGTGCTCTGAGTGAGGATACCGGCACTGCCAGTAAACTCACCTTT  GTGTRLQVTL   \n",
       "2  TGTGCCAGCAGCCTCCTTACACCGACGGACGACAATGAGCAGTTCTTC   GPGTRLTVL   \n",
       "3        TGTGCCAGCAGCTTAGACGGCACTAACTATGGCTACACCTTC   GSGTRLTVV   \n",
       "4     TGTGTGGTGAACATGGGGGGCCCAGGAACCTACAAATACATCTTT  GTGTRLKVLA   \n",
       "\n",
       "                           fwr4_nt  reads umis raw_clonotype_id  \\\n",
       "0     GGACAAGGCACCAGACTCACAGTTGTAG   2891    3      clonotype69   \n",
       "1  GGGACTGGAACAAGACTTCAGGTCACGCTCG    801    2     clonotype195   \n",
       "2     GGGCCAGGGACACGGCTCACCGTGCTAG  23208   10     clonotype195   \n",
       "3     GGTTCGGGGACCAGGTTAACCGTTGTAG   9293    3      clonotype16   \n",
       "4  GGAACAGGCACCAGGCTGAAGGTTTTAGCAA   6908    3      clonotype88   \n",
       "\n",
       "           raw_consensus_id exact_subclonotype_id  \n",
       "0   clonotype69_consensus_1                     1  \n",
       "1  clonotype195_consensus_2                     1  \n",
       "2  clonotype195_consensus_1                     1  \n",
       "3   clonotype16_consensus_1                     1  \n",
       "4   clonotype88_consensus_3                     1  \n",
       "\n",
       "[5 rows x 31 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "\n",
    "folder_path = './GSE252432_RAW'\n",
    "dataframes = {}\n",
    "\n",
    "# Loop through all .csv.gz files in the folder\n",
    "for file_name in os.listdir(folder_path):\n",
    "    if file_name.endswith('.csv.gz'):\n",
    "        full_path = os.path.join(folder_path, file_name)\n",
    "        df = pd.read_csv(full_path, compression='gzip')\n",
    "        dataframes[file_name] = df  # Store in dict with filename as key\n",
    "\n",
    "# Example: show one of them\n",
    "dataframes['GSM8001307_KH_48_TCR.csv.gz'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9824d61e-d68c-41cb-a846-3aed339900aa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AnnData object with n_obs × n_vars = 390373 × 21906\n",
      "    obs: 'nCount_RNA', 'nFeature_RNA', 'percent.mt', 'predicted.celltype.l2', 'LibraryName', 'PFS_6M', 'Timepoint'\n",
      "    var: 'gene'\n",
      "    obsm: 'X_ref.umap'\n"
     ]
    }
   ],
   "source": [
    "import scanpy as sc\n",
    "adata_all = sc.read_h5ad(\"GSE253173_single_cell_DREAM.h5ad\")\n",
    "print(adata_all)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0f134c03-5c70-400a-b26a-70776211e426",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Based on barcode, match single cell and TCR data\n",
    "from collections import defaultdict\n",
    "\n",
    "def match_barcodes_stop_early(df, adata, threshold=100):\n",
    "    \"\"\"\n",
    "    Iterate over barcodes in df, count matches in adata.obs['LibraryName'],\n",
    "    and stop if any library's count exceeds 'threshold'.\n",
    "    Returns the name of the library that exceeded the threshold, or None\n",
    "    if none exceeded it.\n",
    "    \"\"\"\n",
    "    library_counts = defaultdict(int)\n",
    "\n",
    "    for bc in df[\"barcode\"].unique():\n",
    "        # Remove trailing '-1' if present\n",
    "        pattern = bc.rsplit('-', 1)[0]  \n",
    "        \n",
    "        # Find matching rows in adata.obs based on partial string match\n",
    "        mask = adata.obs.index.str.contains(pattern, regex=False)\n",
    "        matched_obs = adata.obs[mask]\n",
    "        \n",
    "        # Increment counts for each LibraryName\n",
    "        for lib in matched_obs[\"LibraryName\"]:\n",
    "            library_counts[lib] += 1\n",
    "            if library_counts[lib] > threshold:\n",
    "                # As soon as a single library's count passes threshold, return it\n",
    "                return lib\n",
    "\n",
    "    # If we never exceeded 'threshold', return None\n",
    "    return None\n",
    "\n",
    "\n",
    "tcr_file_map = {}\n",
    "for file_name, df in dataframes.items():\n",
    "    max_lib = match_barcodes_stop_early(df, adata_all, threshold=100)\n",
    "    if max_lib is not None:\n",
    "        tcr_file_map[max_lib] = file_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c5e15bde-bd5f-43b5-8422-aea4e265cf10",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "KH_1 GSM8001307_KH_48_TCR.csv.gz\n",
      "KH_10 GSM8001317_KH_58_TCR.csv.gz\n",
      "KH_11 GSM8001318_KH_59_TCR.csv.gz\n",
      "KH_12 GSM8001319_KH_60_TCR.csv.gz\n",
      "KH_13 GSM8001320_KH_61_TCR.csv.gz\n",
      "KH_14 GSM8001321_KH_62_TCR.csv.gz\n",
      "KH_15 GSM8001322_KH_63_TCR.csv.gz\n",
      "KH_2 GSM8001308_KH_49_TCR.csv.gz\n",
      "KH_26 GSM8001323_KH_64_TCR.csv.gz\n",
      "KH_27 GSM8001324_KH_65_TCR.csv.gz\n",
      "KH_28 GSM8001325_KH_66_TCR.csv.gz\n",
      "KH_29 GSM8001326_KH_67_TCR.csv.gz\n",
      "KH_3 GSM8001309_KH_50_TCR.csv.gz\n",
      "KH_30 GSM8001327_KH_68_TCR.csv.gz\n",
      "KH_31 GSM8001328_KH_69_TCR.csv.gz\n",
      "KH_32 GSM8001330_KH_71_TCR.csv.gz\n",
      "KH_33 GSM8001331_KH_72_TCR.csv.gz\n",
      "KH_34 GSM8001332_KH_73_TCR.csv.gz\n",
      "KH_35 GSM8001333_KH_74_TCR.csv.gz\n",
      "KH_36 GSM8001334_KH_75_TCR.csv.gz\n",
      "KH_37 GSM8001335_KH_76_TCR.csv.gz\n",
      "KH_38 GSM8001336_KH_77_TCR.csv.gz\n",
      "KH_39 GSM8001337_KH_78_TCR.csv.gz\n",
      "KH_4 GSM8001311_KH_52_TCR.csv.gz\n",
      "KH_40 GSM8001338_KH_79_TCR.csv.gz\n",
      "KH_41 GSM8001355_KH_96_TCR.csv.gz\n",
      "KH_42 GSM8001356_KH_97_TCR.csv.gz\n",
      "KH_43 GSM8001357_KH_98_TCR.csv.gz\n",
      "KH_44 GSM8001310_KH_51_TCR.csv.gz\n",
      "KH_46 GSM8001358_KH_99_TCR.csv.gz\n",
      "KH_47 GSM8001329_KH_70_TCR.csv.gz\n",
      "KH_5 GSM8001312_KH_53_TCR.csv.gz\n",
      "KH_6 GSM8001313_KH_54_TCR.csv.gz\n",
      "KH_7 GSM8001314_KH_55_TCR.csv.gz\n",
      "KH_8 GSM8001315_KH_56_TCR.csv.gz\n",
      "KH_9 GSM8001316_KH_57_TCR.csv.gz\n",
      "MJ_162 GSM8001375_YY116_TCR.csv.gz\n",
      "MJ_163 GSM8001376_YY117_TCR.csv.gz\n",
      "MJ_164 GSM8001377_YY118_TCR.csv.gz\n",
      "MJ_165 GSM8001378_YY119_TCR.csv.gz\n",
      "MJ_166 GSM8001379_YY120_TCR.csv.gz\n",
      "MJ_167 GSM8001380_YY121_TCR.csv.gz\n",
      "MJ_168 GSM8001381_YY122_TCR.csv.gz\n",
      "MJ_169 GSM8001382_YY123_TCR.csv.gz\n",
      "MJ_170 GSM8001383_YY124_TCR.csv.gz\n",
      "MJ_171 GSM8001384_YY125_TCR.csv.gz\n",
      "MJ_172 GSM8001385_YY126_TCR.csv.gz\n",
      "MJ_173 GSM8001386_YY127_TCR.csv.gz\n",
      "MJ_174 GSM8001387_YY128_TCR.csv.gz\n",
      "MJ_175 GSM8001388_YY129_TCR.csv.gz\n",
      "MJ_176 GSM8001389_YY130_TCR.csv.gz\n",
      "MJ_177 GSM8001390_YY131_TCR.csv.gz\n",
      "MJ_178 GSM8001339_KH_80_TCR.csv.gz\n",
      "MJ_179 GSM8001340_KH_81_TCR.csv.gz\n",
      "MJ_180 GSM8001341_KH_82_TCR.csv.gz\n",
      "MJ_181 GSM8001342_KH_83_TCR.csv.gz\n",
      "MJ_182 GSM8001343_KH_84_TCR.csv.gz\n",
      "MJ_183 GSM8001344_KH_85_TCR.csv.gz\n",
      "MJ_184 GSM8001345_KH_86_TCR.csv.gz\n",
      "MJ_185 GSM8001346_KH_87_TCR.csv.gz\n",
      "MJ_186 GSM8001347_KH_88_TCR.csv.gz\n",
      "MJ_187 GSM8001348_KH_89_TCR.csv.gz\n",
      "MJ_188 GSM8001349_KH_90_TCR.csv.gz\n",
      "MJ_189 GSM8001350_KH_91_TCR.csv.gz\n",
      "MJ_190 GSM8001351_KH_92_TCR.csv.gz\n",
      "MJ_191 GSM8001352_KH_93_TCR.csv.gz\n",
      "MJ_192 GSM8001353_KH_94_TCR.csv.gz\n",
      "MJ_193 GSM8001354_KH_95_TCR.csv.gz\n",
      "YY_38 GSM8001391_YY84_TCR.csv.gz\n",
      "YY_39 GSM8001392_YY85_TCR.csv.gz\n",
      "YY_40 GSM8001393_YY86_TCR.csv.gz\n",
      "YY_41 GSM8001394_YY87_TCR.csv.gz\n",
      "YY_42 GSM8001395_YY88_TCR.csv.gz\n",
      "YY_43 GSM8001396_YY89_TCR.csv.gz\n",
      "YY_44 GSM8001397_YY90_TCR.csv.gz\n",
      "YY_45 GSM8001398_YY91_TCR.csv.gz\n",
      "YY_46 GSM8001399_YY92_TCR.csv.gz\n",
      "YY_47 GSM8001400_YY93_TCR.csv.gz\n",
      "YY_48 GSM8001401_YY94_TCR.csv.gz\n",
      "YY_49 GSM8001402_YY95_TCR.csv.gz\n",
      "YY_50 GSM8001403_YY96_TCR.csv.gz\n",
      "YY_51 GSM8001404_YY97_TCR.csv.gz\n",
      "YY_52 GSM8001405_YY98_TCR.csv.gz\n",
      "YY_53 GSM8001406_YY99_TCR.csv.gz\n",
      "YY64 GSM8001359_YY100_TCR.csv.gz\n",
      "YY65 GSM8001360_YY101_TCR.csv.gz\n",
      "YY66 GSM8001361_YY102_TCR.csv.gz\n",
      "YY67 GSM8001362_YY103_TCR.csv.gz\n",
      "YY68 GSM8001363_YY104_TCR.csv.gz\n",
      "YY69 GSM8001364_YY105_TCR.csv.gz\n",
      "YY70 GSM8001365_YY106_TCR.csv.gz\n",
      "YY71 GSM8001366_YY107_TCR.csv.gz\n",
      "YY72 GSM8001367_YY108_TCR.csv.gz\n",
      "YY73 GSM8001368_YY109_TCR.csv.gz\n",
      "YY74 GSM8001369_YY110_TCR.csv.gz\n",
      "YY75 GSM8001370_YY111_TCR.csv.gz\n",
      "YY76 GSM8001371_YY112_TCR.csv.gz\n",
      "YY77 GSM8001372_YY113_TCR.csv.gz\n",
      "YY78 GSM8001373_YY114_TCR.csv.gz\n",
      "YY79 GSM8001374_YY115_TCR.csv.gz\n"
     ]
    }
   ],
   "source": [
    "for sample in adata_all.obs['LibraryName'].unique():\n",
    "    print(sample,tcr_file_map[sample])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "95caf253-2a8f-47d6-9a03-6cc70c068a44",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "# Save the dictionary to a file\n",
    "with open('tcr_file_map.pkl', 'wb') as f:\n",
    "    pickle.dump(tcr_file_map, f)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
